File: app/sensor_data.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from math import isfinite
import re
from typing import Any, Mapping, Sequence


DEVICE_STATUSES = frozenset({"active", "disabled"})
OBSERVATION_QUALITIES = frozenset({"good", "suspect"})
CANONICAL_UNITS = {
    "air_temperature": "celsius",
    "soil_temperature": "celsius",
    "air_humidity": "percent",
    "soil_moisture": "percent",
    "rainfall": "millimeter",
    "battery": "percent",
    "water_level": "millimeter",
    "pressure": "hectopascal",
    "wind_speed": "meter_per_second",
    "soil_ec": "microsiemens_per_cm",
    "soil_ph": "ph",
}
_MACHINE_ID = re.compile(r"^[a-z][a-z0-9_.-]{0,63}$")
_UNIT_ID = re.compile(r"^[a-z][a-z0-9_./-]{0,31}$")
_UTC_SECOND = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$")


def _required(value: object, label: str) -> str:
    text = str(value or "").strip()
    if not text:
        raise ValueError(f"{label} is required")
    return text


def _optional(value: object) -> str:
    return str(value or "").strip()


def _machine_id(value: object, label: str, *, unit: bool = False) -> str:
    text = _required(value, label)
    pattern = _UNIT_ID if unit else _MACHINE_ID
    if not pattern.fullmatch(text):
        raise ValueError(f"{label} must be a stable lowercase machine identifier")
    return text


def _utc_timestamp(value: object, label: str) -> str:
    text = _required(value, label)
    if not _UTC_SECOND.fullmatch(text):
        raise ValueError(f"{label} must be UTC YYYY-MM-DDTHH:MM:SSZ")
    try:
        parsed = datetime.strptime(text, "%Y-%m-%dT%H:%M:%SZ")
    except ValueError as exc:
        raise ValueError(f"{label} must be UTC YYYY-MM-DDTHH:MM:SSZ") from exc
    if parsed.strftime("%Y-%m-%dT%H:%M:%SZ") != text:
        raise ValueError(f"{label} must be UTC YYYY-MM-DDTHH:MM:SSZ")
    return text


def _number(value: object, label: str) -> float:
    if isinstance(value, bool):
        raise ValueError(f"{label} must be numeric")
    try:
        number = float(value)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"{label} must be numeric") from exc
    if not isfinite(number):
        raise ValueError(f"{label} must be finite")
    return number


@dataclass(frozen=True)
class SensorDevice:
    id: str
    name: str
    provider: str
    field_id: str = ""
    external_id: str = ""
    status: str = "active"
    notes: str = ""

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> "SensorDevice":
        return cls(
            id=_required(value.get("id"), "device id"),
            name=_required(value.get("name"), "device name"),
            provider=_required(value.get("provider"), "provider"),
            field_id=_optional(value.get("field_id")),
            external_id=_optional(value.get("external_id")),
            status=_required(value.get("status") or "active", "device status"),
            notes=_optional(value.get("notes")),
        )

    def validate(self) -> None:
        _required(self.id, "device id")
        _required(self.name, "device name")
        _required(self.provider, "provider")
        if self.status not in DEVICE_STATUSES:
            raise ValueError(f"Unsupported device status: {self.status}")


@dataclass(frozen=True)
class SensorChannel:
    id: str
    device_id: str
    metric: str
    unit: str
    label: str = ""

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> "SensorChannel":
        return cls(
            id=_required(value.get("id"), "channel id"),
            device_id=_required(value.get("device_id"), "device id"),
            metric=_machine_id(value.get("metric"), "metric"),
            unit=_machine_id(value.get("unit"), "unit", unit=True),
            label=_optional(value.get("label")),
        )

    def validate(self) -> None:
        _required(self.id, "channel id")
        _required(self.device_id, "device id")
        metric = _machine_id(self.metric, "metric")
        unit = _machine_id(self.unit, "unit", unit=True)
        expected = CANONICAL_UNITS.get(metric)
        if expected is not None and unit != expected:
            raise ValueError(f"Metric {metric} requires canonical unit {expected}")
        if expected is None and not metric.startswith("custom."):
            raise ValueError("Unknown metric must use the custom. namespace")


@dataclass(frozen=True)
class SensorObservation:
    id: str
    channel_id: str
    observed_at: str
    value: float
    quality: str = "good"
    source_ref: str = ""

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> "SensorObservation":
        return cls(
            id=_required(value.get("id"), "observation id"),
            channel_id=_required(value.get("channel_id"), "channel id"),
            observed_at=_utc_timestamp(value.get("observed_at"), "observed_at"),
            value=_number(value.get("value"), "observation value"),
            quality=_required(value.get("quality") or "good", "observation quality"),
            source_ref=_optional(value.get("source_ref")),
        )

    def validate(self) -> None:
        _required(self.id, "observation id")
        _required(self.channel_id, "channel id")
        _utc_timestamp(self.observed_at, "observed_at")
        _number(self.value, "observation value")
        if self.quality not in OBSERVATION_QUALITIES:
            raise ValueError(f"Unsupported observation quality: {self.quality}")


def project_device(
    device: SensorDevice,
    channels: Sequence[SensorChannel],
    observations: Sequence[SensorObservation],
) -> dict[str, object]:
    """Normalize one sensor/API device into deterministic channel snapshots.

    This is deliberately transport-free: network credentials, endpoints and polling
    state do not belong to the shared observation contract.
    """

    device.validate()
    channel_by_id: dict[str, SensorChannel] = {}
    for channel in channels:
        channel.validate()
        if channel.device_id != device.id:
            raise ValueError("sensor channel belongs to another device")
        if channel.id in channel_by_id:
            raise ValueError(f"Duplicate sensor channel id: {channel.id}")
        channel_by_id[channel.id] = channel

    observation_ids: set[str] = set()
    grouped: dict[str, list[SensorObservation]] = {key: [] for key in channel_by_id}
    for observation in observations:
        observation.validate()
        if observation.id in observation_ids:
            raise ValueError(f"Duplicate sensor observation id: {observation.id}")
        observation_ids.add(observation.id)
        if observation.channel_id not in channel_by_id:
            raise ValueError("sensor observation references an unknown channel")
        grouped[observation.channel_id].append(observation)

    channel_snapshots: list[dict[str, object]] = []
    for channel_id in sorted(channel_by_id):
        channel = channel_by_id[channel_id]
        ordered = sorted(
            grouped[channel_id], key=lambda item: (item.observed_at, item.id)
        )
        latest = ordered[-1] if ordered else None
        channel_snapshots.append(
            {
                "channel_id": channel.id,
                "metric": channel.metric,
                "unit": channel.unit,
                "label": channel.label,
                "latest_value": None if latest is None else latest.value,
                "latest_quality": "" if latest is None else latest.quality,
                "latest_observed_at": "" if latest is None else latest.observed_at,
                "latest_source_ref": "" if latest is None else latest.source_ref,
                "observation_count": len(ordered),
            }
        )

    return {
        "device_id": device.id,
        "field_id": device.field_id,
        "name": device.name,
        "provider": device.provider,
        "external_id": device.external_id,
        "status": device.status,
        "notes": device.notes,
        "channels": channel_snapshots,
    }

File: android/app/src/main/java/gr/mastixa/manager/SensorData.java
package gr.mastixa.manager;

import java.time.Instant;
import java.time.format.DateTimeParseException;
import java.util.*;
import java.util.regex.Pattern;

/** Pure Phase 15 sensor/API normalization contract. No database or network dependency. */
public final class SensorData {
    public static final Set<String> DEVICE_STATUSES = Set.of("active", "disabled");
    public static final Set<String> OBSERVATION_QUALITIES = Set.of("good", "suspect");
    public static final Map<String, String> CANONICAL_UNITS = Map.ofEntries(
        Map.entry("air_temperature", "celsius"),
        Map.entry("soil_temperature", "celsius"),
        Map.entry("air_humidity", "percent"),
        Map.entry("soil_moisture", "percent"),
        Map.entry("rainfall", "millimeter"),
        Map.entry("battery", "percent"),
        Map.entry("water_level", "millimeter"),
        Map.entry("pressure", "hectopascal"),
        Map.entry("wind_speed", "meter_per_second"),
        Map.entry("soil_ec", "microsiemens_per_cm"),
        Map.entry("soil_ph", "ph")
    );

    private static final Pattern MACHINE_ID = Pattern.compile("^[a-z][a-z0-9_.-]{0,63}$");
    private static final Pattern UNIT_ID = Pattern.compile("^[a-z][a-z0-9_./-]{0,31}$");
    private static final Pattern UTC_SECOND = Pattern.compile(
        "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}Z$"
    );

    private SensorData() {}

    public record Device(
        String id,
        String name,
        String provider,
        String fieldId,
        String externalId,
        String status,
        String notes
    ) {}

    public record Channel(
        String id,
        String deviceId,
        String metric,
        String unit,
        String label
    ) {}

    public record Observation(
        String id,
        String channelId,
        String observedAt,
        double value,
        String quality,
        String sourceRef
    ) {}

    public record ChannelSnapshot(
        String channelId,
        String metric,
        String unit,
        String label,
        Double latestValue,
        String latestQuality,
        String latestObservedAt,
        String latestSourceRef,
        int observationCount
    ) {}

    public record Snapshot(
        String deviceId,
        String fieldId,
        String name,
        String provider,
        String externalId,
        String status,
        String notes,
        List<ChannelSnapshot> channels
    ) {}

    private static String text(String value) { return value == null ? "" : value.trim(); }

    private static String required(String value, String label) {
        String normalized = text(value);
        if (normalized.isEmpty()) throw new IllegalArgumentException(label + " is required");
        return normalized;
    }

    private static String machineId(String value, String label, boolean unit) {
        String normalized = required(value, label);
        Pattern pattern = unit ? UNIT_ID : MACHINE_ID;
        if (!pattern.matcher(normalized).matches())
            throw new IllegalArgumentException(label + " must be a stable lowercase machine identifier");
        return normalized;
    }

    private static String utcTimestamp(String value, String label) {
        String normalized = required(value, label);
        if (!UTC_SECOND.matcher(normalized).matches())
            throw new IllegalArgumentException(label + " must be UTC YYYY-MM-DDTHH:MM:SSZ");
        try {
            Instant parsed = Instant.parse(normalized);
            if (!parsed.toString().equals(normalized))
                throw new IllegalArgumentException(label + " must be UTC YYYY-MM-DDTHH:MM:SSZ");
        } catch (DateTimeParseException error) {
            throw new IllegalArgumentException(label + " must be UTC YYYY-MM-DDTHH:MM:SSZ", error);
        }
        return normalized;
    }

    private static void validateDevice(Device device) {
        Objects.requireNonNull(device, "device");
        required(device.id(), "device id");
        required(device.name(), "device name");
        required(device.provider(), "provider");
        if (!DEVICE_STATUSES.contains(device.status()))
            throw new IllegalArgumentException("Unsupported device status: " + device.status());
    }

    private static void validateChannel(Channel channel) {
        Objects.requireNonNull(channel, "channel");
        required(channel.id(), "channel id");
        required(channel.deviceId(), "device id");
        String metric = machineId(channel.metric(), "metric", false);
        String unit = machineId(channel.unit(), "unit", true);
        String expected = CANONICAL_UNITS.get(metric);
        if (expected != null && !expected.equals(unit))
            throw new IllegalArgumentException(
                "Metric " + metric + " requires canonical unit " + expected
            );
        if (expected == null && !metric.startsWith("custom."))
            throw new IllegalArgumentException("Unknown metric must use the custom. namespace");
    }

    private static void validateObservation(Observation observation) {
        Objects.requireNonNull(observation, "observation");
        required(observation.id(), "observation id");
        required(observation.channelId(), "channel id");
        utcTimestamp(observation.observedAt(), "observed_at");
        if (!Double.isFinite(observation.value()))
            throw new IllegalArgumentException("observation value must be finite");
        if (!OBSERVATION_QUALITIES.contains(observation.quality()))
            throw new IllegalArgumentException(
                "Unsupported observation quality: " + observation.quality()
            );
    }

    public static Snapshot project(
        Device device,
        List<Channel> channels,
        List<Observation> observations
    ) {
        validateDevice(device);
        Objects.requireNonNull(channels, "channels");
        Objects.requireNonNull(observations, "observations");

        Map<String, Channel> channelById = new HashMap<>();
        for (Channel channel : channels) {
            validateChannel(channel);
            if (!channel.deviceId().equals(device.id()))
                throw new IllegalArgumentException("sensor channel belongs to another device");
            if (channelById.putIfAbsent(channel.id(), channel) != null)
                throw new IllegalArgumentException("Duplicate sensor channel id: " + channel.id());
        }

        Set<String> observationIds = new HashSet<>();
        Map<String, List<Observation>> grouped = new HashMap<>();
        for (String channelId : channelById.keySet()) grouped.put(channelId, new ArrayList<>());
        for (Observation observation : observations) {
            validateObservation(observation);
            if (!observationIds.add(observation.id()))
                throw new IllegalArgumentException(
                    "Duplicate sensor observation id: " + observation.id()
                );
            List<Observation> bucket = grouped.get(observation.channelId());
            if (bucket == null)
                throw new IllegalArgumentException("sensor observation references an unknown channel");
            bucket.add(observation);
        }

        List<String> channelIds = new ArrayList<>(channelById.keySet());
        Collections.sort(channelIds);
        List<ChannelSnapshot> projected = new ArrayList<>();
        Comparator<Observation> order = Comparator
            .comparing(Observation::observedAt)
            .thenComparing(Observation::id);
        for (String channelId : channelIds) {
            Channel channel = channelById.get(channelId);
            List<Observation> ordered = new ArrayList<>(grouped.get(channelId));
            ordered.sort(order);
            Observation latest = ordered.isEmpty() ? null : ordered.get(ordered.size() - 1);
            projected.add(new ChannelSnapshot(
                channel.id(),
                channel.metric(),
                channel.unit(),
                text(channel.label()),
                latest == null ? null : latest.value(),
                latest == null ? "" : latest.quality(),
                latest == null ? "" : latest.observedAt(),
                latest == null ? "" : text(latest.sourceRef()),
                ordered.size()
            ));
        }

        return new Snapshot(
            device.id(),
            text(device.fieldId()),
            device.name(),
            device.provider(),
            text(device.externalId()),
            device.status(),
            text(device.notes()),
            List.copyOf(projected)
        );
    }
}
